src/Service/GdsService.php

```diff
@@ -0,0 +1,372 @@
+                                "v202:DatumTijdTotmet" => $now->format('Y-m-d\TH:i:s.v\Z'),
+                            ],
+                            "v202:NogNietGerapporteerd" => $notYetReported,
+                            "v202:Sortering"            => [
+                                "v202:Kolom"    => "DATUM_AANMELDING",
+                                "v202:Volgorde" => "DESC",
+                            ],
+                        ],
+                    ],
+                ],
+            ],
+        ];
+
+    }//end createRequestMessage()
+
+
+    /**
+     * Determines if an array is associative.
+     *
+     * @param array $array The array to check.
+     *
+     * @return bool Whether the array is associative.
+     */
+    private function isAssociative(array $array): bool
+    {
+        if ($array === []) {
+            return false;
+        }
+
+        return array_keys($array) !== range(0, (count($array) - 1));
+
+    }//end isAssociative()
+
+
+    /**
+     * Get urls for data from the response from GDS2.
+     *
+     * @param array $result The GDS2 response.
+     *
+     * @return array
+     */
+    private function getDataUrls(array $result): array
+    {
+        $locations = [];
+        $baseUrls  = $result['SOAP-ENV:Body']['ns9:BestandenlijstOpvragenResponse']['ns9:antwoord']['ns3:BaseURLSet']['ns4:BaseURL'];
+
+        foreach ($baseUrls as $baseUrl) {
+            if ($baseUrl['@type'] === 'certificaat') {
+                break;
+            }
+        }
+
+        $files = $result['SOAP-ENV:Body']['ns9:BestandenlijstOpvragenResponse']['ns9:antwoord']['ns5:BestandenLijst']['ns5:Afgifte'];
+
+        if ($this->isAssociative($files) === true) {
+            $fileUrl     = $files['ns8:digikoppeling-external-datareferences']['ns8:data-reference']['ns8:transport']['ns8:location']['ns8:senderUrl']['#'];
+            $mime        = $files['ns8:digikoppeling-external-datareferences']['ns8:data-reference']['ns8:content']['@contentType'];
+            $locations[] = [
+                'url'  => "{$baseUrl['#']}/$fileUrl",
+                'mime' => "$mime",
+            ];
+
+            return $locations;
+        }
+
+        foreach ($files as $file) {
+            $fileUrl     = $file['ns8:digikoppeling-external-datareferences']['ns8:data-reference']['ns8:transport']['ns8:location']['ns8:senderUrl']['#'];
+            $mime        = $file['ns8:digikoppeling-external-datareferences']['ns8:data-reference']['ns8:content']['@contentType'];
+            $locations[] = [
+                'url'  => "{$baseUrl['#']}/$fileUrl",
+                'mime' => "$mime",
+            ];
+        }
+
+        return $locations;
+
+    }//end getDataUrls()
+
+
+    /**
+     * Creates snapshots in the database and dispatch an event to process them.
+     *
+     * @param array $file The contents of a file.
+     *
+     * @return array The created snapshots.
+     */
+    public function createSnapShots(array $file): array
```

Comment:
(This is a big function)

Reply:
It is big, but not very breakable.

Reply:
true, i am just pointing out the obvious here, it would be cleaner code if it was smaller